Svc/ComQueue/ComQueue.cpp

```diff
@@ -0,0 +1,308 @@
+        // Both READY and unknown states should not be possible at this point. To receive a status message we must be
+        // one of the WAITING or RETRY states.
+        default:
+            FW_ASSERT(0, this->m_state);
+            break;
+    }
+}
+
+void ComQueue::run_handler(const NATIVE_INT_TYPE portNum, NATIVE_UINT_TYPE context) {
+    // Downlink the high-water marks for the Fw::ComBuffer array types
+    ComQueueDepth comQueueDepth;
+    for (FwSizeType i = 0; i < comQueueDepth.SIZE; i++) {
+        comQueueDepth[i] = this->m_queues[i].get_high_water_mark();
+        this->m_queues[i].clear_high_water_mark();
+    }
+    this->tlmWrite_comQueueDepth(comQueueDepth);
+
+    // Downlink the high-water marks for the Fw::Buffer array types
+    BuffQueueDepth buffQueueDepth;
+    for (FwSizeType i = 0; i < buffQueueDepth.SIZE; i++) {
+        buffQueueDepth[i] = this->m_queues[i + COM_PORT_COUNT].get_high_water_mark();
+        this->m_queues[i + COM_PORT_COUNT].clear_high_water_mark();
+    }
+    this->tlmWrite_buffQueueDepth(buffQueueDepth);
+}
+
+void ComQueue::retryReturn_handler(const NATIVE_INT_TYPE portNum, Fw::Buffer &fwBuffer) {
+    this->m_lock.lock();
+    this->m_bufferRetry = fwBuffer;
+    this->m_lock.unLock();
+}
+
+
+// ----------------------------------------------------------------------
+// Private helper methods
+// ----------------------------------------------------------------------
+
+void ComQueue::enqueue(const FwIndexType queueNum, QueueType queueType, const U8* data, const FwSizeType size) {
+    // Enqueue the given message onto the matching queue. When no space is available then emit the queue overflow event,
+    // set the appropriate throttle, and move on. Will assert if passed a message for a depth 0 queue.
+    const FwSizeType expectedSize = (queueType == QueueType::COM_QUEUE) ? sizeof(Fw::ComBuffer) : sizeof(Fw::Buffer);
+    const FwIndexType portNum = queueNum - ((queueType == QueueType::COM_QUEUE) ? 0 : COM_PORT_COUNT);
+    FW_ASSERT(expectedSize == size, size, expectedSize);
+    FW_ASSERT(portNum >= 0, portNum);
+    Fw::SerializeStatus status = this->m_queues[queueNum].enqueue(data, size);
+    if (status == Fw::FW_SERIALIZE_NO_ROOM_LEFT && !this->m_throttle[queueNum]) {
+        this->log_WARNING_HI_QueueOverflow(queueType, portNum);
+        this->m_throttle[queueNum] = true;
+    }
+    // When the component is already in READY state process the queue to send out the next available message immediately
+    if (this->m_state == READY) {
+        this->processQueue();
+    }
+}
+
+void ComQueue::sendComBuffer(Fw::ComBuffer& comBuffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_comRetry = comBuffer;
+    this->comQueueSend_out(0, comBuffer, 0);
+    this->m_state = WAITING;
+}
+
+void ComQueue::sendBuffer(Fw::Buffer& buffer) {
+    FW_ASSERT(this->m_state == READY);
+    this->m_lock.lock();
+    // Buffer ownership transferred to downstream component. Clear local reference.
```

Comment:
Why are we clearing out the retry buffer here? We may or may not be sending from the retry buffer.

Reply:
My concern is on its face the code is sending buffer and clearing bufferRetry, and there is no obvious relationship between these two variables. What if we hit this code and the retry buffer is not null and buffer is not the retry buffer? Then we will send buffer and clear the retry buffer, resulting in a lost send. Maybe this can't happen, but I think it would be clearer to write the code so it's obvious this can't happen. Or maybe you could add an assertion against this case.

Reply:
I went ahead and cleaned this up.